src/microcontroller/microcontroller.go

```diff
@@ -12,48 +12,121 @@ import (
+	Solenoids   []component.SolenoidConfig `yaml:"solenoids"`
+	Igniters    []component.IgniterConfig  `yaml:"igniters"`
+}
+
+//GetConfig -
+func (m Microcontroller) GetConfig() (config Config) {
+	config.ID = m.ID
+	config.Host = m.Host
+	config.Port = m.Port
+	config.Solenoids = make([]component.SolenoidConfig, len(m.Solenoids))
+	for i, sol := range m.Solenoids {
+		config.Solenoids[i] = sol.GetConfig()
+	}
+	return
+}
+
+//Load -
+func (m *Microcontroller) Load(config Config) {
+	m.ID = config.ID
+	m.Host = config.Host
+	m.Port = config.Port
+	m.Solenoids = make([]component.Solenoid, len(config.Solenoids))
+	for i, sol := range config.Solenoids {
+		m.Solenoids[i].Load(sol)
+	}
+	m.Igniters = make([]component.Igniter, len(config.Igniters))
+	for i, igniter := range config.Igniters {
+		m.Igniters[i].Load(igniter)
+	}
 }
 
 /*String Just for pretty printing the Microcontroller info */
-func (m *Microcontroller) String() string {
-	return utilities.LabelString("Microcontroller", utilities.LabelString("Id", strconv.Itoa(m.ID))+utilities.LabelString("Host", m.Host)+utilities.LabelString("Port", m.Port)+utilities.LabelString("Solenoids", m.solenoidsString()))
+func (m Microcontroller) String() string {
+	return utilities.LabelString("Microcontroller",
+		utilities.LabelString("Id", strconv.Itoa(m.ID))+
+			utilities.LabelString("Host", m.Host)+
+			utilities.LabelString("Port", m.Port)+
+			utilities.LabelString("Components", m.ComponentString()))
 }
 
 //Init - Initialize all the components on this micro
-func (m *Microcontroller) Init() error {
-	for i := 0; i < len(m.Solenoids); i++ {
-		err := m.Solenoids[i].Init()
-		if err != nil {
-			log.Println("Failed to load: ", m.Solenoids[i].String())
-			return err
+func (m *Microcontroller) Init() (err error) {
+	for _, component := range m.GetComponentMap() {
+		if err = component.Init(); err != nil {
+			log.Println("Failed to load: ", component.String())
+			return
 		}
 	}
-	return nil
+	return
+}
+
+//GetComponentMap - retrieve a map of pointers to all components on this micro
+func (m *Microcontroller) GetComponentMap() map[string]component.Component {
+	count := len(m.Solenoids) + len(m.Igniters)
+	components := make(map[string]component.Component, count)
+	for i := 0; i < len(m.Solenoids); i++ {
+		components["Solenoid_"+strconv.Itoa(i)] = &m.Solenoids[i]
+	}
+	for i := 0; i < len(m.Igniters); i++ {
+		components["Igniter_"+strconv.Itoa(i)] = &m.Igniters[i]
+	}
+	return components
+}
+
+//NewMicrocontroller -
+func NewMicrocontroller(host string, port string) (Microcontroller, error) {
+	micro := Microcontroller{
+		Host: host,
+		Port: port,
+	}
+	err := micro.LoadConfigFromFile()
+	if err != nil {
+		return Microcontroller{}, err
+	}
+	return micro, nil
+
 }
 
 //ToFullAddress returns a network address including the ip address and port that this micro is listening on
-func (m *Microcontroller) ToFullAddress() string {
+func (m Microcontroller) ToFullAddress() string {
```

Comment:
curious why this isnt being referenced as a pointer

Reply:
The pointer set is only for methods that modify the struct or call functions with side effects that would modify the struct. This is just using data out of the struct and therefore should go into the base set.